src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncRewriter.AsyncIteratorRewriter.cs

```diff
@@ -101,17 +99,18 @@ protected override void GenerateControlFields()
                     boolType,
                     GeneratedNames.MakeAsyncIteratorPromiseIsActiveFieldName(), isPublic: true);
 
+                // the element type may contain method type parameters, which are now alpha-renamed into type parameters of the generated class
+                TypeSymbol elementType = ((AsyncStateMachine)stateMachineType).ElementType;
```

Comment:
Did we consider parameterizing the base type on the state machine type?

Reply:
You mean to avoid the cast? I could look into that

---
In reply to: [221788308](https://github.com/dotnet/roslyn/pull/30105#discussion_r221788308) [](ancestors = 221788308)